### PR TITLE
Fix FlatList item generic types

### PR DESCRIPTION
**Pull request by jkadamczyk**
## Summary

I think I ran into a typescript types regression, where the `Animated.FlatList` type was no longer generic. It always used the any type and I couldn't provide a custom item type myself anymore. I found it while upgrading from Reanimated v2 to v3. I think I managed to patch the generated .d.ts file properly locally and this is what did the trick for me.

If this isn't matching your code style or requirements, I'm happy to work with you on making this work.

## Test plan

Probably makes sense to see if it breaks anything in the example app but other than that 🤷

### CODE REVIEW

File: src/reanimated2/component/FlatList.tsx [hunk 1]
```diff
@@ -60,48 +60,58 @@ interface AnimatedFlatListComplement<T> extends FlatList<T> {
...
+export const ReanimatedFlatList = forwardRef(FlatListForwardRefRender) as <
+  // We need explicit any here, because this is the exact same type that is used in React Native types.
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  ItemT = any
+>(
+  props: ReanimatedFlatListPropsWithLayout<ItemT> & {
+    ref?: ForwardedRef<FlatList>;
   }
-);
+) => React.ReactElement;
```
Comment (tjzel) [inline src/reanimated2/component/FlatList.tsx:114]: This makes me curious, are we certain that this type of general type here won't cause any issues?

**Review by tjzel — CHANGES_REQUESTED** [review]
Thank you for finding that @jkadamczyk! I took the liberty and committed some minor changes directly into your PR. Everything looks well, apart from that one question that I asked. Once you assure me everything is fine I'm happy to merge it 😃

File: src/reanimated2/component/FlatList.tsx (hunk 1 above)
Comment (jkadamczyk) [reply src/reanimated2/component/FlatList.tsx:114]: Hey! Let me look into that!

Comment (jkadamczyk) [reply src/reanimated2/component/FlatList.tsx:113]: @tjzel I wasn't able to find a more elegant solution. If you have some ideas, please share them with me.
To back using it like it is I can share the function call type declaration in the ExoticComponent interface. It also returns a `ReactElement`. But for sure there are parts of that missing.
CleanShot 2023-12-28 at 19 06 51@2x

In my solution which I have in a patch already I was using below sources:
- https://fettblog.eu/typescript-react-generic-forward-refs/
- https://stackoverflow.com/questions/58469229/react-with-typescript-generics-while-using-react-forwardref

Comment (jkadamczyk) [reply src/reanimated2/component/FlatList.tsx:113]: Also https://github.com/Microsoft/TypeScript/issues/17574

Comment (tjzel) [reply src/reanimated2/component/FlatList.tsx:113]: Hey, that's some amazing research there, very interesting reads! (it's tempting to request more changes just to read some more 😄)

Let's merge this bad boy

**Review by tjzel — APPROVED** [review]

**Pull request merged**